junit-jupiter-migration-support/src/main/java/org/junit/jupiter/migrationsupport/rules/ParameterizedExtension.java

```diff
@@ -0,0 +1,266 @@
+
+	/**
+	 * Indicate whether we can provide parameterized support.
+	 * This requires the testClass to either have a static {@code @Parameters} method
+	 * and correct {@code @Parameter} and their corresponding values
+	 * or to have a constructor that could be injected.
+	 */
+	@Override
+	public boolean supports(ContainerExtensionContext context) {
+		return hasParametersMethod(context) && hasCorrectParameterFields(context);
+	}
+
+	@Override
+	public Iterator<TestTemplateInvocationContext> provide(ContainerExtensionContext context) {
+		//grabbing the parent ensures the paremeters are stored in the same store.
+		return context.getParent().flatMap(ParameterizedExtension::parameters).map(
+			ParameterizedExtension::testTemplateContextsFromParameters).orElse(Collections.emptyIterator());
+	}
+
+	/**
+	 * Since the parameterized runner in JUnit 4 could only resolve constructor parameters
+	 * this extension once again here only support them on the constructor and require an {@code @Parameters} method
+	 *
+	 * @param parameterContext the context for the parameter to be resolved; never
+	 * {@code null}
+	 * @param extensionContext the extension context for the {@code Executable}
+	 * about to be invoked; never {@code null}
+	 * @return true if the above is met otherwise false.
+	 */
+	@Override
+	public boolean supports(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		return hasParametersMethod(extensionContext)
+				&& parameterContext.getDeclaringExecutable() instanceof Constructor;
+	}
+
+	@Override
+	public Object resolve(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		int parameterCount = parameterContext.getDeclaringExecutable().getParameterCount();
+		Object[] parameters = resolveParametersForConstructor(extensionContext, parameterCount);
+
+		int parameterIndex = parameterContext.getIndex();
+		//move to the next set of parametersFields
+		if (lastParameterToBeResolved(parameterContext)) {
+			this.parameterCount.incrementAndGet();
+		}
+
+		return parameters[parameterIndex];
+	}
+
+	/**
+	 * Retrieves the Object[] of the current iteration we are working on.
+	 *
+	 * @param extensionContext the extensionContext
+	 * @param parameterCount the amount of parameters of the constructor.
+	 *
+	 * @return the object[] for this parameter iteration.
+	 * @throws ParameterResolutionException If the amount of arguments of the constructor doesn't match the amount
+	 * of arguments of the currently resolved object[]
+	 */
+	private Object[] resolveParametersForConstructor(ExtensionContext extensionContext, int parameterCount)
+			throws ParameterResolutionException {
+		return parameters(extensionContext).map(ArrayList::new).map(l -> l.get(this.parameterCount.get())).filter(
+			params -> params.length == parameterCount).orElseThrow(
+				ParameterizedExtension::unMatchedAmountOfParametersException);
```

Comment:
I've had to put quite some effort to understand the code in this class because of the extensive use of lambdas. It is probably me... I am not smart enough or young enough to understand these(... well definitely both) :)

I was especially troubled by the above lambda chain. I think it would have helped me if `this.parameterCount` was named differently. In this case it has the same name as the `int parameterCount`. I think a more descriptive name would be something like `parametersCollectionSize` or `parameterArraysCount`